Implement QgsGeometry.as_numpy and .as_shapely

merydian opened the pull request:
## Description

Adds methods to QgsGeometry to convert it to a (list of) numpy array(s) or shapely object(s). This is part of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/227. This enhancement improves integration with NumPy and Shapely while keeping them an optional dependency.

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

merydian committed:
Implement QgsGeometry.as_numpy and .as_shapely

rouault commented on python/PyQt6/core/__init__.py.in:
```diff
@@ -540,6 +540,7 @@ QgsException.__doc__ = "Defines a QGIS exception class."
 
 try:
    import numpy as _numpy
+   import shapely as _shapely
```
shapely related things should be put in another try block, so that the raster part can work only if numpy is available

merydian committed:
Place shapely related things in separate try statement

roya0045 commented:
What about geometries with Z/M coordinates?

merydian committed:
Remove unnecessary import

merydian committed:
Improve performance of as_shapely

By utilizing QgsGeometry.asWkb() this is ~8.45 times faster than the previous implementation.

nyalldawson commented on python/PyQt6/core/__init__.py.in:
```diff
@@ -598,6 +625,33 @@ except ModuleNotFoundError:
+   def _geometry_as_shapely(self) -> _shapely.geometry.base.BaseGeometry:
+      wkb_qbytearray = self.asWkb()  # Get the geometry in WKB format (QByteArray)
+      wkb_bytes = bytes(wkb_qbytearray)
+      shapely_geom = _shapely.from_wkb(wkb_bytes)
```
Does this work? It would save a copy of the data, and might squeeze out a bit more performance:

```suggestion
      shapely_geom = _shapely.from_wkb(wkb_qbytearray.data())
```

merydian committed:
Use QByteArray.data() for as_shapely

wonder-sk commented:
@merydian  I am now getting a runtime error on QGIS start:
```
Couldn't load SIP module.  
Python support will be disabled.

Traceback (most recent call last):  
  File "", line 1, in  
  File "/home/martin/qgis/git-master/build/output/python/qgis/__init__.py", line 86, in  
    import qgis.gui  
  File "/home/martin/qgis/git-master/build/output/python/qgis/gui/__init__.py", line 26, in  
    from qgis.core import Qgis as _Qgis  
  File "/home/martin/qgis/git-master/build/output/python/qgis/core/__init__.py", line 22754, in  
    def _geometry_as_shapely(self) -> _shapely.geometry.base.BaseGeometry:  
AttributeError: module 'shapely' has no attribute 'geometry'
```

This is on ubuntu 22.04, shapely 1.8.0

I was able to fix this by doing:
```python
import shapely.geometry as _sg
```
and then rewriting the offending line:
```python
def _geometry_as_shapely(self) -> _sg.base.BaseGeometry:
```

Not sure why that does not work for me, and whether the fix is correct, but I assume others may get a similar issue...

wonder-sk commented:
For the record, on ubuntu 24.04, shapely 2.0.3 things work fine without any modifications :slightly_smiling_face:

merydian commented:
Might this be a `shapely` related issue? 

There seem to be other people having [similar issues](https://stackoverflow.com/questions/35978649/python-shapely-install-not-working). I think your fix is simple enough to just include? Tagging @nyalldawson for some input please.
